module/zfs/zio.c

```diff
@@ -654,6 +654,37 @@ zio_add_child(zio_t *pio, zio_t *cio)
 	mutex_exit(&pio->io_lock);
 }
 
+void
+zio_add_child_first(zio_t *pio, zio_t *cio)
+{
+	/*
+	 * Logical I/Os can have logical, gang, or vdev children.
+	 * Gang I/Os can have gang or vdev children.
+	 * Vdev I/Os can only have vdev children.
+	 * The following ASSERT captures all of these constraints.
+	 */
+	ASSERT3S(cio->io_child_type, <=, pio->io_child_type);
+
+	zio_link_t *zl = kmem_cache_alloc(zio_link_cache, KM_SLEEP);
+	zl->zl_parent = pio;
+	zl->zl_child = cio;
+
+	ASSERT(list_is_empty(&cio->io_parent_list));
+	list_insert_head(&cio->io_parent_list, zl);
+
+	mutex_enter(&pio->io_lock);
+
+	ASSERT(pio->io_state[ZIO_WAIT_DONE] == 0);
+
+	uint64_t *countp = pio->io_children[cio->io_child_type];
+	for (int w = 0; w < ZIO_WAIT_TYPES; w++)
+		countp[w] += !cio->io_state[w];
+
+	list_insert_head(&pio->io_child_list, zl);
+
```

Comment:
I am just wondering if `pio->io_child_count` and `cio->io_parent_count` should be incremented in here like in `zoo_add_child()`?

Reply:
They would have to be, but I removed them all together in ccec7fbe1c6. :)

Reply:
Geez oh crap. I was looking at an older branch of my O_DIRECT code just now when I cross referencing your changes lol... Sorry about that. This all looks good to me.